### ISSUE TITLE
Spike: what happens to in-progress Celery tasks when their EC2 instance is terminated?

### ISSUE THREAD

**Issue by seanh** [hypothesis/lms#5241]
### Question

Elastic Beanstalk terminates EC2 instances all the time: during deployments, autoscaling down, etc. What happens if a Celery worker is in the middle of executing a task when the EC2 instance that it's running on gets terminated?

### Answer

Detailed notes from me investigating this are in the comments below, but I've updated this issue description with a summary of the answer:

* If the Celery task uses [early acknowledgment](https://docs.celeryq.dev/en/stable/glossary.html#term-early-acknowledgment) (which is Celery's default) then Celery will have acknowledged the message to RabbitMQ removing the message from the queue before ever calling the user-provided task function. If the worker then gets terminated in the middle of executing the task function then **the task will never be completed**.

* If the Celery task uses [late acknowledgment](https://docs.celeryq.dev/en/stable/glossary.html#term-late-acknowledgment) (which I'd recommend) then Celery won't acknowledge the message to RabbitMQ until _after_ the user-provided task function returns. If the worker gets terminated in the middle of executing the task function then RabbitMQ will notice this and will **immediately re-deliver the task to another worker**. This means that **the task may get executed multiple times** so you should make sure that your task is idempotent or that you can tolerate "occasional" task duplication.

  I say "immediately" re-deliver because there appears to be a **12s** delay in re-delivery. The RabbitMQ developers say this delay shouldn't exist and we think it could be a phantom delay due to messages missing from our logs but we aren't sure. See <https://github.com/hypothesis/lms/issues/5241#issuecomment-1513247645> for details.

  I say "occasional" task duplication because we actually have no idea how frequently this happens but I suspect it's not too often.

* **A task may get delivered to another worker that is also about to be terminated** as happened in my experiment in <https://github.com/hypothesis/lms/issues/5241#issuecomment-1513247645>. This can result in the task being executed more than twice and in the possibly phantom **12s** delay being multiplied.

**Comment by seanh** [comment]
What happens when Celery task is killed by <kbd>Ctrl</kbd> + <kbd>c</kbd> in the local dev env?

Using this Celery task to test with:

```diff
diff --git a/lms/tasks/email_digests.py b/lms/tasks/email_digests.py
index c4df9e427..142505a43 100644
--- a/lms/tasks/email_digests.py
+++ b/lms/tasks/email_digests.py
@@ -1,4 +1,5 @@
 import logging
+import time
 from datetime import datetime, timedelta, timezone
 from typing import List, Optional
 
@@ -11,6 +12,13 @@ from lms.tasks.celery import app
 LOG = logging.getLogger(__name__)
 
 
+@app.task
+def sleep(seconds):
+    LOG.info("Sleeping")
+    time.sleep(seconds)
+    LOG.info("Finished sleeping")
+
+
 @app.task
 def send_instructor_email_digest_tasks(batch_size):
     """
```

To test run `make dev` in one terminal and in another terminal run `make shell` and run the task manually:

```python
$ make shell
...
>>> from lms.tasks.email_digests import sleep
>>> sleep.delay(100)
<AsyncResult: 8fe8d183-6ed4-456d-9407-e6029e20fa0d>
>>> sleep.delay(100)
<AsyncResult: f5f46713-e93e-4b70-ae08-769ccb101873>
>>> 
```

**Comment by seanh** [comment]
There is a relevant difference between [supervisord-dev.conf](https://github.com/hypothesis/lms/blob/main/conf/supervisord-dev.conf#L25-L26) (used in the development environment) and [supervisord.conf](https://github.com/hypothesis/lms/blob/main/conf/supervisord.conf) (used in production): `supervisord-dev.conf` has `stopsignal = KILL` and `stopasgroup = true`:

https://github.com/hypothesis/lms/blob/b31e0efbbe92970b2ce4e1781360d33f29d5f3a1/conf/supervisord-dev.conf#L25-L26

The `stopsignal` setting controls the signal used to kill the program when a stop is requested, usually `TERM`, `HUP`, `INT`, `QUIT`, `KILL`, `USR1`, or `USR2` (default: `TERM`).

`stopasgroup` (default: `false`) causes Supervisor to send the `stopsignal` to the whole process group and implies `killasgroup`. It's meant for preventing orphaned child processes when parent processes don't propagate stop signals to their children.

There's also `stopwaitsecs`: the number of seconds to wait after signalling a process to stop before killing it, default: `10`.

See [Supervisor's config documentation](http://supervisord.org/configuration.html).

**Comment by seanh** [comment]
### `acks_late=True` is the culprit

We have `acks_late=True` in our Celery config:

https://github.com/hypothesis/lms/blob/b31e0efbbe92970b2ce4e1781360d33f29d5f3a1/lms/tasks/celery.py#L33-L34

See Celery's [docs for `acks_late`](https://docs.celeryq.dev/en/stable/userguide/tasks.html?highlight=acks_late#Task.acks_late) and also the [Should I use retry or acks_late?](https://docs.celeryq.dev/en/stable/faq.html?highlight=acks_late#should-i-use-retry-or-acks-late) FAQ.

If I change this to `acks_late=False` then the Celery worker does give the task time to finish before terminating.

I don't see anything in Celery's [docs about stopping workers](https://docs.celeryq.dev/en/stable/userguide/workers.html#stopping-the-worker) that says that `acks_late=True` causes in-progress tasks to be terminated immediately but it makes sense: Celery's programmers are probably assuming that if your task has `acks_late=True` then it's idempotent and so it's okay to kill the in-progress task immediately and restart it later.

**Comment by seanh** [comment]
### Docker

Whether or not the Docker layer is going to be an issue here depends on what Elastic Beanstalk does to stop a Docker container. Based on the presence of the [`docker stop`](https://docs.docker.com/engine/reference/commandline/stop/) command in `/opt/elasticbeanstalk/config/private/eb-docker-stop` on our Elastic Beanstalk instances @robertknight guessed that it might stop Docker containers by running `docker stop`.

If so we may have a problem: `docker stop` sends `TERM` to Supervisor and then after 10s sends `KILL`. You can increase the timeout with `docker stop --time` but that requires Elastic Beanstalk to provide us with a way to specify a value that should be passed to the `--time` argument.

I've tested this locally using `make run-docker`: `docker stop <container>` gives a running task 10s then kills it. `docker stop --time 100 <container>` give it 100s.

**Comment by seanh** [comment]
### Elastic Beanstalk kills Supervisor and Celery workers immediately

[Slack thread](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1680604108756809).

I added a test Celery task to https://github.com/hypothesis/test-pyramid-app and we deployed the app to https://test-pyramid-app.hypothes.is/. Based on testing the production deployment of the test app it [looks like](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1680800937822729?thread_ts=1680604108.756809&cid=C4K6M7P5E) Elastic Beanstalk kills the Docker container's entry point process immediately without giving Supervisor or the Celery worker process any chance to do a graceful or "warm" shutdown and give any in-progress Celery tasks any time to complete.

As noted in the Slack comment this seems like it will have been causing an unknown number of emails sent by h to not get sent (I don't think h currently retries its email-sending tasks if they fail).

**Comment by seanh** [comment]
### What does RabbitMQ do when a Celery worker gets killed mid-task?

To answer this I ran an experiment using https://github.com/hypothesis/test-pyramid-app/ in [production](https://test-pyramid-app.hypothes.is/): I started off a Celery task that would take 30 mins (1800 secs) to complete (using test-pyramid-app's [Celery task that counts down from 1800](https://github.com/hypothesis/test-pyramid-app/blob/28a9699169c9ee45b186e686966cc3204335a022/test_pyramid_app/celery.py#L12-L17) logging the current number one second at a time as it counts down) then I used GitHub Actions to kick off a redeploy while the task was executing.

TLDR: When the Celery worker is terminated RabbitMQ "immediately" re-delivers the task to another worker which starts executing the task from the start again. RabbitMQ knows as soon as the Celery worker has been terminated because it closes the TCP connection between the worker and RabbitMQ. I say "immediately" in quotes because there's a `~12s` delay between the last log message from the first Celery worker while it was executing the task and the first log message from the second Celery worker that the task was re-delivered to, but according to the RabbitMQ developers ([I asked them in their Slack](https://rabbitmq.slack.com/archives/C1EDN83PA/p1681824210831719?thread_ts=1681813132.202369&cid=C1EDN83PA)) this delay shouldn't exist and the re-delivery should be immediate. It could be that something somewhere in our systems is introducing a delay or [it could be due to missing log messages](https://hypothes-is.slack.com/archives/CR3E3S7K8/p1681824635969789), we don't know.

But in my experiment **the task was re-delivered to another worker that was about to be terminated!** The app had two EC2 instances that were both in the process of being terminated and replaced by two new instances. When the worker on the first instance lost its connection RabbitMQ re-delivered the task to the worker on the second instance which briefly started processing the task and then was also terminated and then RabbitMQ re-delivered the task again to a worker on one of the new instances. This results in a total `24s` delay between the last log message from the first worker and the first log message from the third worker which would finally complete the task uninterrupted.

Logs:

The test-pyramid-app's two instances before the redeploy were **i-05ffa96cab7b83768** and **i-07ff3966332d1ab5b**. These two instances were terminated by the deploy. The deploy replaced these with two new instances **i-0182bc01ffc1fa927** and **i-02d34bb1f4cc2ff00**.

1. At **13:46:15** a worker on **i-05ffa96cab7b83768** received task instance **87033bbb-e486-4079-a7c5-b92eb5d27ae7** and began counting from **1800**:

       2023-04-18T12:46:15.000Z test-pyramid-app-prod_i-05ffa96cab7b83768 eb-c9e46f476ac4-stdouterr.log worker (stderr)      | [2023-04-18 12:46:15,363: INFO/MainProcess] Task test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7] received
       2023-04-18T12:46:15.000Z test-pyramid-app-prod_i-05ffa96cab7b83768 eb-c9e46f476ac4-stdouterr.log worker (stderr)      | [2023-04-18 12:46:15,371: INFO/ForkPoolWorker-2] test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7]: Working 1800

2. I started a redeploy at **13:46:58**

3. The worker seemed to continue until **13:54:05** by which time it had counted down to **1330**:

       2023-04-18T12:54:05.000Z test-pyramid-app-prod_i-05ffa96cab7b83768 eb-c9e46f476ac4-stdouterr.log worker (stderr)      | [2023-04-18 12:54:05,966: INFO/ForkPoolWorker-2] test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7]: Working 1330

4. At **13:54:17** (12 seconds after the last message logged by the original worker) a second worker on **i-07ff3966332d1ab5b** received the same task (**87033bbb-e486-4079-a7c5-b92eb5d27ae7**) and started counting again from **1800**:

       2023-04-18T12:54:17.000Z test-pyramid-app-prod_i-07ff3966332d1ab5b eb-54ca44cf27ae-stdouterr.log worker (stderr)      | [2023-04-18 12:54:17,570: INFO/MainProcess] Task test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7] received
       2023-04-18T12:54:17.000Z test-pyramid-app-prod_i-07ff3966332d1ab5b eb-54ca44cf27ae-stdouterr.log worker (stderr)      | [2023-04-18 12:54:17,570: INFO/ForkPoolWorker-2] test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7]: Working 1800

   This one log message (**1800**) seems to be the only message logged by this worker.

5. At **13:54:29** (24 seconds after the last message logged by the first worker, 12 seconds after the last message logged by the second worker) the same task (**87033bbb-e486-4079-a7c5-b92eb5d27ae7**) is received by a **third** worker on **i-02d34bb1f4cc2ff00** which begins counting from **1800** for a third time:

       2023-04-18T12:54:29.000Z test-pyramid-app-prod_i-02d34bb1f4cc2ff00 eb-74b1a980b2a8-stdouterr.log worker (stderr)      | [2023-04-18 12:54:29,061: INFO/MainProcess] Task test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7] received
       2023-04-18T12:54:29.000Z test-pyramid-app-prod_i-02d34bb1f4cc2ff00 eb-74b1a980b2a8-stdouterr.log worker (stderr)      | [2023-04-18 12:54:29,063: INFO/ForkPoolWorker-2] test_pyramid_app.celery.work[87033bbb-e486-4079-a7c5-b92eb5d27ae7]: Working 1800

   This third worker continued uninterrupted until the task was finished.
